### PR TITLE
Clean up and optimize BasisUniversal

### PR DESCRIPTION
**Pull request by BlueCube3310**
This PR aims to clean up BasisUniversal's code to make it easier to read, as well as extracts the code from `register_types.cpp` into its own files.

It also slightly optimizes the memory usage and compression times by using the `get_mipmap_offset_size_and_dimensions()` function instead of copying each mipmap's data into a new image.

While it seems to function correctly, it still needs some testing, especially on mobile devices.

### CODE REVIEW

**Comment by fire** [comment]
I'm generally ok with this but I agree that testing on mobile devices is required.

**Comment by akien-mga** [comment]
Unlike our usual practice, such a change would be much easier to review if done as two separate commits: one to move the code to a separate file, and one to actual do logic changes. With both types of changes made in the same commit it's pretty difficult to review.

**Comment by BlueCube3310** [comment]
The commits are now split, as suggested.

File: modules/basis_universal/image_compress_basisu.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	int basisu_err = compressor.process();
+	ERR_FAIL_COND_V(basisu_err != basisu::basis_compressor::cECSuccess, basisu_data);
+
+	const basisu::uint8_vec &basisu_out = compressor.get_output_basis_file();
+	basisu_data.resize(basisu_out.size() + 4);
+
+	// Copy the encoded data to the buffer.
+	{
+		uint8_t *w = basisu_data.ptrw();
+		(*(uint32_t *)w) = decompress_format;
```
Comment (AThousandShips) [inline modules/basis_universal/image_compress_basisu.cpp:136]:
```suggestion
		*(uint32_t *)w = decompress_format;
```

**Review by akien-mga — APPROVED** [review]
Looks really good!

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**